src/Microsoft.Extensions.Configuration.AzureAppConfiguration/Extensions/ConfigurationClientExtensions.cs

```diff
@@ -65,131 +64,47 @@ public static async Task<KeyValueChange> GetKeyValueChange(this ConfigurationCli
-            }
+            List<MatchConditions> newMatchConditions = new List<MatchConditions>();
 
-            if (keyValues.Any(k => !string.Equals(k.Label.NormalizeNull(), options.Label.NormalizeNull())))
-            {
-                throw new ArgumentException("All key-values registered for refresh must use the same label.", $"{nameof(keyValues)}[].{nameof(ConfigurationSetting.Label)}");
-            }
+            AsyncPageable<ConfigurationSetting> pageable = client.GetConfigurationSettingsAsync(selector, cancellationToken);
 
-            if (keyValues.Any(k => k.Label != null && k.Label.Contains("*")))
-            {
-                throw new ArgumentException("The label filter cannot contain '*'", $"{nameof(options)}.{nameof(options.Label)}");
-            }
+            using IEnumerator<MatchConditions> existingMatchConditionsEnumerator = matchConditions.GetEnumerator();
 
-            var hasKeyValueCollectionChanged = false;
-            var selector = new SettingSelector
+            await foreach (Page<ConfigurationSetting> page in pageableManager.GetPages(pageable, matchConditions).ConfigureAwait(false))
             {
-                KeyFilter = options.KeyFilter,
-                LabelFilter = string.IsNullOrEmpty(options.Label) ? LabelFilter.Null : options.Label,
-                Fields = SettingFields.ETag | SettingFields.Key
-            };
+                Response response = page.GetRawResponse();
 
-            // Dictionary of eTags that we write to and use for comparison
-            var eTagMap = keyValues.ToDictionary(kv => kv.Key, kv => kv.ETag);
+                newMatchConditions.Add(new MatchConditions { IfNoneMatch = response.Headers.ETag });
 
-            // Fetch e-tags for prefixed key-values that can be used to detect changes
-            await TracingUtils.CallWithRequestTracing(options.RequestTracingEnabled, RequestType.Watch, options.RequestTracingOptions,
-                async () =>
+                if (response.Status == (int)HttpStatusCode.OK &&
+                    (!existingMatchConditionsEnumerator.MoveNext() || !existingMatchConditionsEnumerator.Current.IfNoneMatch.Equals(response.Headers.ETag)))
                 {
-                    await foreach (ConfigurationSetting setting in client.GetConfigurationSettingsAsync(selector, cancellationToken).ConfigureAwait(false))
-                    {
-                        if (!eTagMap.TryGetValue(setting.Key, out ETag etag) || !etag.Equals(setting.ETag))
-                        {
-                            hasKeyValueCollectionChanged = true;
-                            break;
-                        }
-
-                        eTagMap.Remove(setting.Key);
-                    }
-                }).ConfigureAwait(false);
-
-            // Check for any deletions
-            if (eTagMap.Any())
-            {
-                hasKeyValueCollectionChanged = true;
+                    hasCollectionChanged = true;
```

Comment:
Shall we break the for loop here?

Reply:
I didn't break because I assumed we need to create the whole `newMatchConditions` object even if we see that there's a change, so we need to continue querying all the pages to get all of the new page etags. I could change something about the condition/add a comment to make it more clear.